### PR TITLE
replace not found data nil with an ErrNotFound

### LINKED ISSUE #1543: Panic during performance test
running performance test: networks/redis on current master

```
time="2022-10-12T09:48:38Z" level=error msg="IssueVC failed" error="unable to publish the issued credential: failed to publish credential, error while creating transaction: unable to add newly created transaction to State: unable to obtain Redis transaction-level write lock: Database Error: redsync: failed to acquire lock" module=VCR operation=IssueVC operationID=IssueVC requestURI=/internal/vcr/v2/issuer/vc user="<nil>"
time="2022-10-12T09:48:38Z" level=error msg="IssueVC failed" error="unable to publish the issued credential: failed to publish credential, error while creating transaction: unable to add newly created transaction to State: unable to obtain Redis transaction-level write lock: Database Error: 1 error occurred:\n\t* context deadline exceeded\n\n" module=VCR operation=IssueVC operationID=IssueVC requestURI=/internal/vcr/v2/issuer/vc user="<nil>"
echo: http: panic serving 172.21.0.11:49920: runtime error: invalid memory address or nil pointer dereference
goroutine 679510 [running]:
net/http.(*conn).serve.func1()
        /usr/local/go/src/net/http/server.go:1850 +0xbf
panic({0x1577180, 0x2436eb0})
        /usr/local/go/src/runtime/panic.go:890 +0x262
github.com/nuts-foundation/nuts-node/network.(*Network).calculateLamportClock(0xc0001a2300, {0x1a27ec0, 0xc000044030}, {0xc0150bc300?, 0x2, 0x1a27ec0?})
        /go/network/network.go:544 +0xfd
github.com/nuts-foundation/nuts-node/network.(*Network).CreateTransaction(0xc0001a2300, {{0x1a27280, 0xc012dafa60}, {0xc0200c3200, 0x477, 0x480}, {0x17736b7, 0x13}, 0x0, {0x32cb50b8, ...}, ...})
        /go/network/network.go:497 +0xa2b
github.com/nuts-foundation/nuts-node/vcr/issuer.networkPublisher.PublishCredential({{_, _}, {_, _}, {_, _}, {_, _}}, {{0xc014802900, 0x3, ...}, ...}, ...)
        /go/vcr/issuer/network_publisher.go:101 +0x595
github.com/nuts-foundation/nuts-node/vcr/issuer.issuer.Issue({{_, _}, {_, _}, {_, _}, _, {_, _}}, {{0xc014d30480, ...}, ...}, ...)
        /go/vcr/issuer/issuer.go:92 +0x4bd
github.com/nuts-foundation/nuts-node/vcr/api/v2.Wrapper.IssueVC({{0x0, 0x0}, {0x1a1d8a0, 0xc000297320}, {0x1a2ffb8, 0xc0002a4b40}}, {0x1a3eff8, 0xc011ab5720})
        /go/vcr/api/v2/api.go:138 +0x30b
github.com/nuts-foundation/nuts-node/vcr/api/v2.(*ServerInterfaceWrapper).IssueVC(0xc000305f60, {0x1a3eff8, 0xc011ab5720})
        /go/vcr/api/v2/generated.go:1932 +0x8f
github.com/nuts-foundation/nuts-node/vcr/api/v2.RegisterHandlersWithBaseURL.func2({0x1a3eff8, 0xc011ab5720})
        /go/vcr/api/v2/generated.go:2143 +0x7f
github.com/labstack/echo/v4.(*Echo).add.func1({0x1a3eff8, 0xc011ab5720})
        /pkg/mod/github.com/labstack/echo/v4@v4.9.0/echo.go:536 +0x51
github.com/labstack/echo-contrib/prometheus.(*Prometheus).HandlerFunc.func1({0x1a3eff8, 0xc011ab5720})
        /pkg/mod/github.com/labstack/echo-contrib@v0.13.0/prometheus/prometheus.go:406 +0x230
github.com/labstack/echo/v4/middleware.RequestLoggerConfig.ToMiddleware.func1.1({0x1a3eff8, 0xc011ab5720})
        /pkg/mod/github.com/labstack/echo/v4@v4.9.0/middleware/request_logger.go:219 +0x168
github.com/nuts-foundation/nuts-node/http.decodeURIPath.func1({0x1a3eff8, 0xc011ab5720})
        /go/http/engine.go:212 +0xd6
github.com/labstack/echo/v4.(*Echo).ServeHTTP(0xc0000cdb00, {0x1a26b60?, 0xc021e09340}, 0xc008d0ca00)
        /pkg/mod/github.com/labstack/echo/v4@v4.9.0/echo.go:646 +0x3d1
net/http.serverHandler.ServeHTTP({0x1a23b40?}, {0x1a26b60, 0xc021e09340}, 0xc008d0ca00)
        /usr/local/go/src/net/http/server.go:2947 +0x30c
net/http.(*conn).serve(0xc011ee7c20, {0x1a27f30, 0xc000846e40})
        /usr/local/go/src/net/http/server.go:1991 +0x607
created by net/http.(*Server).Serve
        /usr/local/go/src/net/http/server.go:3102 +0x4db
time="2022-10-12T09:48:38Z" level=warning msg="Conn has unread data" module=Storage
time="2022-10-12T09:48:38Z" level=error msg="IssueVC failed" error="unable to publish the issued credential: failed to publish credential, error while creating transaction: unable to add newly created transaction to State: unable to obtain Redis transaction-level write lock: Database Error: redsync: failed to acquire lock" module=VCR operation=IssueVC operationID=IssueVC requestURI=/internal/vcr/v2/issuer/vc user="<nil>"
```

### PR DESCRIPTION
**Pull request by gerardsn**
closes #1543

`dag.getTransaction` and `dag.readPayload` now return `dag.ErrTransactionNotFound` and `dag.ErrPayloadNotFound` when the data is not found.

### CODE REVIEW

File: network/api/v1/api.go [hunk 1]
```diff
@@ -65,11 +66,11 @@ func (a Wrapper) GetTransaction(ctx echo.Context, hashAsString string) error {
 	}
 	transaction, err := a.Service.GetTransaction(hash)
 	if err != nil {
+		if errors.Is(err, dag.ErrTransactionNotFound) {
+			return core.NotFoundError("transaction not found")
```
Comment (pascaldekloe) [inline network/api/v1/api.go:70]: Bit of a vague description. Which transaction are we talking about?

The `err`or is discarded here. What message did it tell?

At this point we have nothing to add to the error (context), and thus err should be returned. GetTransaction(hash) has all the information required to produce a descriptive message. Specifically, it should return something like `core.NetFoundError("hash %#x lookup: %w", hash, ErrNotFound)` with:

```go
var ErrNotFound = errors.New("no such Nuts transaction")
```

File: network/dag/dag.go [hunk 2]
```diff
@@ -146,7 +147,13 @@ func (d *dag) Migrate() error {
 				for _, ref := range heads {
 					transaction, err := getTransaction(ref, tx)
 					if err != nil {
+						if errors.Is(err, ErrTransactionNotFound) {
+							return fmt.Errorf("database migration failed: head not found (%s=%s)", core.LogFieldTransactionRef, ref)
```
Comment (pascaldekloe) [inline network/dag/dag.go:151]: Again, the error message is lost. Can *never* do that. Either log, return or deal/absorb the error.

**Review by pascaldekloe — CHANGES_REQUESTED** [review]
ErrNotFound over nil returns is a good thing. 👍

File: network/api/v1/api.go (hunk 1 above)
Comment (gerardsn) [reply network/api/v1/api.go:70]: This is an interpretation of the error to generate the correct response for the API call.
`core.NotFoundError` is used to return a statuscode 404 with the entered message.

Comment (pascaldekloe) [reply network/api/v1/api.go:70]: If it's just a classifier then `core.NotFoundError("%w", err)`.

File: network/dag/dag.go (hunk 2 above)
Comment (gerardsn) [reply network/dag/dag.go]: This case should never happen. If this fails the node won't start and the error here provides a more useful debugging info than the returning the original error. (head is a specific transaction)

File: network/api/v1/api.go (hunk 1 above)
Comment (gerardsn) [reply network/api/v1/api.go:70]: That requires an overhaul of the current API error handling. Out of scope for this PR

File: network/dag/dag.go (hunk 2 above)
Comment (gerardsn) [reply network/dag/dag.go]: now wraps the original error

Comment (pascaldekloe) [reply network/dag/dag.go]: You can not make assumptions on how useful an error is. Never! The logic may change over time. All that is required is to preserve and amend with any decision making along the execution chain.

File: network/api/v1/api.go (hunk 1 above)
Comment (pascaldekloe) [reply network/api/v1/api.go:70]: No just that ☝️ instead of `core.NotFoundError("transaction not found")`.

Comment (gerardsn) [reply network/api/v1/api.go:70]: An API should not return an internal error right? Just the fact that the requested item is not found.

Comment (pascaldekloe) [reply network/api/v1/api.go:70]: Correct. APIs should not propagate errors. They log and use status codes.

**Pull request merged**